src/Core/Domain/Address/Command/SetRequiredFieldsForAddressCommand.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Address\Command;
+
+use PrestaShop\PrestaShop\Core\Domain\Address\Exception\InvalidAddressRequiredFieldsException;
+use PrestaShop\PrestaShop\Core\Domain\Address\ValueObject\RequiredFields;
+
+/**
+ * Sets required fields for new address when adding
+ */
+class SetRequiredFieldsForAddressCommand
+{
+    /**
+     * @var string[]
+     */
+    private $requiredFields;
+
+    /**
+     * @param string[] $requiredFields
+     *
+     * @throws InvalidAddressRequiredFieldsException
+     */
+    public function __construct(array $requiredFields)
+    {
+        $this->assertContainsOnlyAllowedFields($requiredFields);
+
+        $this->requiredFields = $requiredFields;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getRequiredFields()
+    {
+        return $this->requiredFields;
+    }
+
+    /**
+     * Check that all provided fields are allowed.
+     *
+     * @param string[] $requiredFields
+     *
+     * @throws InvalidAddressRequiredFieldsException
+     */
+    private function assertContainsOnlyAllowedFields(array $requiredFields)
+    {
+        if (empty($requiredFields)) {
+            return;
+        }
+
+        if (!empty(array_diff($requiredFields, RequiredFields::ALLOWED_REQUIRED_FIELDS))) {
+            throw new InvalidAddressRequiredFieldsException(
+                sprintf(
+                    'Invalid address required fields provided. Allowed fields are: %s',
+                    implode(',', RequiredFields::ALLOWED_REQUIRED_FIELDS)
```

Comment:
@PierreRambaud I know you've played with this part recently 😛 do you think this is a bad thing in the context of shop customization ? This prevents a module developer to add custom fields and make them required, is that an issue ?

Reply:
Yes it's an issue, you'll not be able to add custom field that are only manage by hooks :/

Reply:
There was no explicit hook for required fields so how should we implement this in migration? And is this an issue for this PR @PierreRambaud @matks ?

Reply:
@RaimondasSapola At controller level, modules can add new required fields (using existing hooks). This looks like a relevant usecase so I think we need to address it 🤔

Reply:
Ok i am a bit confused. So the problem here is that module developers cannot add new required fields after migration and we should look into implementing hooks in for requiredFields migration or is it that this validation only allows predefined required fields to be added and module developers might want to add completely new fields?

Reply:
@RaimondasSapola Both 😄

> So the problem here is that module developers cannot add new required fields after migration

Yes because `this validation only allows predefined required fields to be added`

> we should look into implementing hooks in for requiredFields migration

Since this is controlled on controller level and module can override controllers using Symfony re-routing I think it's already OK but I'll check. Leave this to me 👍 

> module developers might want to add completely new fields?

Yes, this is a 100% relevant usecase. Module developers want to add new fields in the database and want to be able to:
- read them
- list them in grids
- modify them in add/edit forms
- make them required fields

Reply:
So in this case validation should be removed altogether?